### PR TITLE
Update ProductsControl to use API search when product catalog is large

### LINKED ISSUE #505: Product(s)Control: Query for products takes a long time with a lot of products
A store owner with about 5000 products reported that the Hand-picked Products block is taking at least 30 seconds to load ([forum post](https://wordpress.org/support/topic/handpicked-products-query-extremely-slow/)).

This is likely due to the pre-loading of all products into the `ProductsControl` component, which fetches products in batches of 100, but waits for all products to return before rendering them. The intention was to load all items into the `SearchListControl`, to make searching very fast (no waiting for a server response).

Let's benchmark how this works on 100, 500, 1000, 5000+ products – once 1000+ products are loaded, is JS searching really faster than querying the API again?

Could we use two methods for these `ProductControl`s, one for smaller stores and one for larger product databases (so we can keep the fast search for the smaller product catalogs)?

Overall, this might benefit from some local caching, so the products don't have to be fetched every time the `ProductControl` is loaded.

### PR DESCRIPTION
**Pull request by ryelle**
Fixes #505 – If a store has more than 200 products, the existing products search list can take 10+ seconds to load, which negates the performance benefit of the existing search. To balance this, we can switch to using pagination and an API search process if the store has 200+ products.

### How to test the changes in this Pull Request:

~Since the `onSearch` prop is not added to the SearchListControl yet, you can test this and see there is no functionality change, or you can tell the component to use the local file, with [this branch checked out.](https://github.com/woocommerce/woocommerce-admin/pull/1992)~ The `@woocommerce/components` package has been published, you can test this properly now. Make sure to run `npm install` before building.

To test, add a Hand-picked Products block (the product list in Featured Products will also get this treatment, if this looks good, but it's not yet). Before you add the block, open the Network tab and watch the XHR requests.

If your store has less than 200 products, you'll only see the API calls when the ProductsControl loads. Searching the list should be quick, and not trigger any other API requests.

If your store has over 200 products, you'll see two API calls when the ProductsControl loads, one for the first page of products, and just one for the selected products. Each search will also trigger two API requests, for your search term, and one for the selected products again, since we need to keep them in the list.

In any case, you should be able to find the products you're searching for.

### CODE REVIEW

File: assets/php/class-wgpb-block-library.php
Comment (coderkevin) [inline assets/php/class-wgpb-block-library.php]: I wonder if it might be better to let the php code decide if it's a "large" catalog or not, and just return a boolean here, rather than have the js code decide.

File: assets/js/components/products-control/index.js
Comment (coderkevin) [inline assets/js/components/products-control/index.js]: Should the 400 here be a constant somewhere to be more descriptive and so it can be used elsewhere?

**Review by coderkevin — APPROVED** [review]
Looks good and tests good, just a couple of optional comments. Pre-approving.

File: assets/js/components/products-control/index.js
Comment (ryelle) [reply assets/js/components/products-control/index.js]: I think it's fine as a plain number, I think it's pretty clear what's going on.

**Comment by timmyc** [comment]
@ryelle did you want another review here, or is this good to go?

**Comment by ryelle** [comment]
I need to publish the `@woocommerce/*` packages before this can be merged, so this is on hold for today (I'll probably be able to publish tomorrow).

**Pull request merged**
